src/objects.c

```diff
@@ -496,39 +496,45 @@ PgDatabase *register_auto_database(const char *name)
 	return db;
 }
 
-/* add or update client users */
-PgGlobalUser *add_global_user(const char *name, const char *passwd)
+static PgGlobalUser *update_global_user_passwd(PgGlobalUser *user, const char *passwd)
 {
-	PgGlobalUser *user = find_global_user(name);
-
-	if (user == NULL) {
-		user = slab_alloc(user_cache);
-		if (!user)
-			return NULL;
-		user->credentials.global_user = user;
-
-		list_init(&user->head);
-		list_init(&user->pool_list);
-		safe_strcpy(user->credentials.name, name, sizeof(user->credentials.name));
-		put_in_order(&user->head, &user_list, cmp_user);
-
-		aatree_insert(&user_tree, (uintptr_t)user->credentials.name, &user->credentials.tree_node);
-		user->pool_mode = POOL_INHERIT;
-		user->pool_size = -1;
-	}
-
+	Assert(user);
 	passwd = passwd ? passwd : "";
 	safe_strcpy(user->credentials.passwd, passwd, sizeof(user->credentials.passwd));
 	user->credentials.dynamic_passwd = strlen(passwd) == 0;
 	return user;
 }
 
-PgCredentials *add_global_credentials(const char *name, const char *passwd)
+static PgGlobalUser *create_new_global_user(const char *name, const char *passwd)
 {
-	PgGlobalUser *user = add_global_user(name, passwd);
+	PgGlobalUser *user = slab_alloc(user_cache);
+
 	if (!user)
 		return NULL;
-	return &user->credentials;
+
+	user->credentials.global_user = user;
+
+	list_init(&user->head);
+	list_init(&user->pool_list);
+	safe_strcpy(user->credentials.name, name, sizeof(user->credentials.name));
+	put_in_order(&user->head, &user_list, cmp_user);
+
+	aatree_insert(&user_tree, (uintptr_t)user->credentials.name, &user->credentials.tree_node);
+	user->pool_mode = POOL_INHERIT;
+	user->pool_size = -1;
+
+	return update_global_user_passwd(user, passwd);
+}
+
+/* add or update client users */
+PgGlobalUser *add_global_user(const char *name, const char *passwd)
```

Comment:
This does basically the same as `find_or_add_global_user`, except that it also updates the password of the global user. 

Let's move these functions closer together and make this difference clear from the naming and function comment.

Reply:
Actually, maybe it's better to completely remove `add_global_user` and explicitely call `update_global_user_passwd` in all places where we need to update the password.